### ISSUE TITLE
Unary RPC received by server after 10 seconds after client side deadline exceeds.

### ISSUE THREAD

**Issue by yyh-graphsql** [grpc/grpc#11725]
Please answer these questions before submitting your issue. 
 
 
### What version of gRPC and what language are you using?
GRPC-1.0.0
C++
 
### What operating system (Linux, Windows, …) and version?
(Ubuntu 4.8.4-2ubuntu1~14.04.3)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3)
 
### What did you do?
I wrote a test case using a single channel which serves for a one bidirectional async stream and other unary RPCs at the same time. The client and server are in the same process. I encountered an unary RPC timeout once. And I found the server side got the unary RPC after 10 seconds, which after the client side got the deadline exceed. I didn't do any failover operations and the resource condition is good.
 
### What did you expect to see?
 All RPCs will success.
 
### What did you see instead?
  I found the server side got a unary RPC after 10 seconds once, which after the client side got the deadline exceed.
 
### Anything else we should know about your project / environment?
One thing is that the unary RPC's timeout was also set 10 seconds in the test case. Pretty like the RPC was sent out after 10 seconds.

**Comment by vjpai** [comment]
Can you provide the code snippet?

**Comment by yyh-graphsql** [comment]
Hi @vjpai 
The following is the MICRO when issuing the unary RPC. The unary RPC's RPCTIMEOUT is 10secs.                                                                                                                                       ```
while (retryMgr.ShouldRetry()) {                                                     \                                                                                                                                                                                            
      gutil::Status connectionRc = validateChannel(RPCTIMEOUT * 1000);                   \
      if (!connectionRc.Ok()) {                                                          \
        rc = connectionRc;                                                               \
        break;                                                                           \
      }                                                                                  \
      grpc::ClientContext context;                                                       \
      {                                                                                  \
        std::chrono::system_clock::time_point deadline =                                 \
          std::chrono::system_clock::now() + std::chrono::milliseconds(RPCTIMEOUT);      \
        context.set_deadline(deadline);                                                  \
        boost::shared_lock<boost::shared_mutex> lock(mutex_);                            \
        status = stub_->RPCWORD(&context, request, &reply);                                     \
      }                                                                                  \
      if (status.ok()) {                                                                 \
        if (reply.rc(reply.rc_size()-1) != gutil::Status::kTimeout) {                    \
          break;                                                                         \
        }                                                                                \
      }                                                                                  \
      else if (status.error_code() == grpc::StatusCode::DEADLINE_EXCEEDED) {             \
        **LOG(ERROR) << "RPC timeout reached, no retry, timeout: " << RPCTIMEOUT;          \**
        rc = gutil::Status::kTimeout;                                                    \
        break;                                                                         \
      }                                                                                  \
      else {                                                                             \
        LOG(ERROR) << "Error found in RPC, code: " << status.error_code()                \
            << ", error msg: " << status.error_message();                                \
        rc = gutil::Status::kIOError;                                                    \
      }                                                                                  \
      LOG(ERROR) << "Retry on error, retry times: " << retryMgr.GetRetryTimes()          \
                  << ", channel state: " << this->GetConnnectionState();                 \
      retryMgr.SleepBeforeRetry();                                                       \
    }
```
In the test I have two kinds of RPC, say 'A' and 'B'. I will issue two 'A's followed with two 'B's each in a batch for 100 times. At the same time in the process's background, there is a bidirectional async stream shared the same channel with the unary RPCs, which client side will issue heartbeat packages to the server side periodically (each per 500ms). In the failed situation, an RPC 'B' was DEADLINE_EXCEEDED and after that, I got three RPC 'B's on the server side. BTW, during the 'B''s timout period, the background async stream worked well.
I still cannot give a minimum reproducible code snippet. Since I also encountered another GRPC problem https://github.com/grpc/grpc/issues/11681, I wonder if I can skip this problem by update our GRPC version.

**Comment by vjpai** [comment]
I can't be 100% sure if this is the issue, but you are holding a lock while doing a synchronous RPC. This, in general, is a bad idea because the lock will be held as long until the sync RPC gets its response. If the async RPCs in your background thread also have to use the same lock before issuing, then they will be locked out. Likewise with any other sync RPCs in other threads. In general, it is not a good idea to hold a lock while waiting for a response form a sync RPC. This is not really an issue with the RPC system; it is an issue with its usage.

We can't help any further without more details about the usage, so I hope this helped. I'll close the issue, but feel free to reopen if you can provide info that shows that my suggestion wasn't helpful.

**Comment by yyh-graphsql** [comment]
@vjpai Thank you for the reminder. The read lock is used here because we will hold a write lock to change the stub_ when switching to the primary. That may block the RPC, but I got `LOG(ERROR) << "RPC timeout reached, no retry, timeout: " << RPCTIMEOUT` in the log which means the RPC really timed out. And in the log, I saw that the RPC showed up on the server side 10s later.

Here are my other concerns:
- Can I fix it by upgrading GRPC if the problem may relate to a known issue?
- Are there any critical c++/c-core bugs in GRPC that make me really need to change the GRPC version? Since when I encountered a problem, it seems I lacked the judgment to figure out if it is a known issue or a GRPC bug. In the release note, the GRPC seems pretty stable for the c core and c++ lang.
